### PR TITLE
[FLINK-3418] Don't run RocksDB copy utils in external process

### PR DESCRIPTION
**Pull request by aljoscha**
This was causing to many problems with security tokens and yarn. Now,
let the RocksDB backup run in a thread but
don't interrupt these Threads anymore on closing. The Threads will close
themselves because the copy operation will fail because of a
FileNotFoundException when the state directories are being cleaned up.

This also removes the ExternalProcessRunner because it is not needed
anymore and using it causes too many headaches.

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java [hunk 1]
```diff
@@ -26,32 +25,46 @@
...
-	}
+		String threadName = "HDFS Copy from " + localPath + " to " + remotePath;
+
+		final List<Exception> asyncException = new ArrayList<>();
+
+		Thread copyThread = new Thread(threadName) {
+			@Override
+			public void run() {
+				try {
+					Configuration hadoopConf = new Configuration();
```
Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java]: This is only loading the Hadoop configuration from the classpath, not from the Flink configuration or environment variables

I think our filesystem code has a method to try the environment variables and the config as well.

Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java]: Ah dammit, I pushed to wrong stuff. Give me a sec.

**Comment by aljoscha** [comment]
Should be the correct code now. :smile:

**Comment by StephanEwen** [comment]
Fails on checkstyle, otherwise looks good...

**Comment by aljoscha** [comment]
Ahh, stupid beginners mistake. Fixed it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java [hunk 2]
```diff
@@ -17,41 +17,50 @@
...
 
-		FileSystem fs = FileSystem.get(new URI(backupUri), hadoopConf);
+	public static void copyFromLocal(final File localPath,
+			final URI remotePath) throws Exception {
+		// Do it in another Thread because HDFS can deadlock if being interrupted while copying
+		String threadName = "HDFS Copy from " + localPath + " to " + remotePath;
 
-		fs.copyFromLocalFile(new Path(localBackupPath), new Path(backupUri));
-	}
+		final List<Exception> asyncException = new ArrayList<>();
```
Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java]: If you have cases like this again in the future: I use Flink's `Tuple1` for those cases. Its probably cheaper than creating a new ArrayList.

**Comment by rmetzger** [comment]
+1 to merge this change

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/util/HDFSCopyFromLocal.java]: Agreed. `ArrayList` will allocate an array of 10 elements when you initialize the ArrayList without an initial capacity. So either initialize the ArrayList with a size of `1` or use Robert's approach.

**Comment by tillrohrmann** [comment]
Thanks for your work @aljoscha. Changes look good to me. I'll address Robert's concern and then merge the PR.

**Pull request closed without merging**